### PR TITLE
Added astylerc and pre-commit hook

### PR DESCRIPTION
**Pull request by matthiasbeyer**
This PR adds a `astylerc` file for formatting with astyle.

I'm not sure whether to add a pre-commit checker or something, too.

### CODE REVIEW

**Comment by neithernut** [comment]
> +--break-blocks

Do we want this?
What about, say:

| int i=8;
| while (i) {
|     //...
|     i>>=1;
| }

**Comment by matthiasbeyer** [comment]
This is a good point. Shall I remove the `--break-blocks` line?

**Comment by neithernut** [comment]
Is there an option which enforces spaces _after_ blocks only?

**Comment by matthiasbeyer** [comment]
Not for astyle, sadly.

**Comment by neithernut** [comment]
Then it should be removed, for now.

**Comment by matthiasbeyer** [comment]
ACK?

**Comment by matthiasbeyer** [comment]
I added a pre-commit hook, which solves #21 . Please review, @neithernut , @TheNeikos !

**Comment by matthiasbeyer** [comment]
If we want to merge this, including the pre-commit hook which was added in this PR, we should do the following:
- Run `astyle` on each file which is in this branch atm, add the changes to _this_ PR
- Rebase all branches onto master after _this_ branch is merged
- Redo all commits in the other PRs, so we only add formatted code

Thoughts?

I would do this ASAP, so we don't get into trouble later on.

**Comment by neithernut** [comment]
ACK.
Astyle-checking might get integrated into the cmake-setup later on, but it's fine for now.

**Comment by matthiasbeyer** [comment]
Have you tested it? I tested it but maybe another test run would be fine!

If you don't want to mess up, just create a new clone of some C-code repository and run the hook when test-committing something.

**Comment by neithernut** [comment]
I'll do this after I've processes all the other notifications in my inbox.

**Comment by neithernut** [comment]
Mh... on this machine, bash is in /bin, not in /usr/bin

**Comment by matthiasbeyer** [comment]
Is there a unified path for bash somewhere? So we can get the shebang line "platform independent" somehow?

**Comment by neithernut** [comment]
I would have to read up on the POSIX standart, but sh was supposed to go into /bin (or be symlink'ed), I think. The most sane thing would be to port the script to sh.

**Comment by matthiasbeyer** [comment]
@manuelmessner Do you want to port the script to (pure) sh?

**Comment by manuelmessner** [comment]
> Is there a unified path for bash somewhere? So we can get the shebang line "platform independent" somehow?

/bin is the standard path. On arch linux bash is in /usr/bin but /bin is linked to /usr/bin due to compatibilty reasons.

> @manuelmessner Do you want to port the script to (pure) sh?

Yes i can do it.

**Comment by matthiasbeyer** [comment]
So then please do it. Contact me directly for workflow questions.

**Comment by matthiasbeyer** [comment]
@neithernut please review & test again. Issues should be fixed now.

File: hooks/pre-commit [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+for file in $(git diff --name-only HEAD | grep "\.[ch]$")
+do
+    echo_cyan "Checking: $file"
+    fpath=$tmpdir/$(basename $file)
+    fpath_formatted=$tmpdir/$(basename $file)-formatted
+
+    cp $file $fpath
+
+    astyle --options=./etc/astylerc < $fpath > $fpath_formatted
```
Comment (neithernut) [inline hooks/pre-commit:49]: Seems like this line produces a newline at the end of the output file.
Running diff afterwards yields a non-null result...

Comment (matthiasbeyer) [inline hooks/pre-commit:49]: @manuelmessner do you want to solve this issue?

**Comment by matthiasbeyer** [comment]
@neithernut Test again, please. I tested and the latest patches solve the issue you mentioned (https://github.com/matthiasbeyer/libreset/pull/10#discussion_r16911344) for me.

File: hooks/pre-commit [hunk 2]
```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+
```
Comment (neithernut) [inline hooks/pre-commit:2]: The script currently only works when committing from the repo's root directory.
To enhance this, I would inset a "cd" to the path returned by "git rev-parse --show-toplevel" here.

Comment (matthiasbeyer) [inline hooks/pre-commit:2]: I wouldn't be sure about this. Did you actually test it? I would bet that git hooks are always executed from the repositories root directory!

Comment (neithernut) [inline hooks/pre-commit:2]: I tested the script directly.
In the past, I _had_ problems with commit-hooks _not_ being executed from the repo's root. So, even if this behaviour is fixed in new versions of git, including it doesn't hurt.

File: hooks/pre-commit [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+rm -rf $tmpdir
+
+if [ $bad -eq 1 ]
+then
+    echo_red "Bad formatted file found. You cannot commit"
+    cd $p
+    exit 1
+fi
+
+cd $p
```
Comment (neithernut) [inline hooks/pre-commit:75]: Should not be neccessary, but it doesn't hurt either.

**Comment by neithernut** [comment]
Seems to work just fine now.

**Comment by matthiasbeyer** [comment]
So I can merge this?

**Comment by neithernut** [comment]
Yep, but I think you should squash a few things first.

**Comment by TheNeikos** [comment]
Why not use the bash builtins `pushd` and `popd` ?

**Comment by matthiasbeyer** [comment]
@neithernut Rebased. I moved the "run astyle on <file>" commits to the end of the branch. As we do not have more `[ch]` files then the two mentioned, this PR should be ready now.

edit: Just noticed: Github does not show the history as-is. The "run 'astyle'..." commits are acutally at the end of the history!

**Comment by matthiasbeyer** [comment]
@TheNeikos Feel free to submit another PR (which can be based on master after the merge).

**Comment by TheNeikos** [comment]
Alright

**Comment by matthiasbeyer** [comment]
For documentation: Closes #21

**Comment by neithernut** [comment]
ACK

**Comment by manuelmessner** [comment]
> Why not use the bash builtins pushd and popd?

These bulitins are not available in pure shells. 
And this would be the first compatiblity problem in the script.

**Pull request merged**